wasm/src/main/scala/ir2wasm/WasmExpressionBuilder.scala

```diff
@@ -1514,4 +1550,93 @@ private class WasmExpressionBuilder private (
     instrs += CALL(FuncIdx(WasmFunctionName.jsLinkingInfo))
     IRTypes.AnyType
   }
+
+  // ===============================================================================
+  // array
+  // ===============================================================================
+  private def genArrayLength(t: IRTrees.ArrayLength): IRTypes.Type = {
+    genTreeAuto(t.array)
+    instrs += ARRAY_LEN
+    IRTypes.IntType
+  }
+
+  private def genNewArray(t: IRTrees.NewArray): IRTypes.Type = {
+    if (t.lengths.isEmpty || t.lengths.sizeIs > t.typeRef.dimensions)
+      throw new AssertionError(
+        s"invalid lengths ${t.lengths} for array type ${t.typeRef.displayName}"
+      )
+
+    def genWithDimensions(typeRef: IRTypes.ArrayTypeRef, lengths: List[IRTrees.Tree]): Unit = {
+      val elemTy = extractArrayElemType(t.typeRef)
+      val wasmElemTy = TypeTransformer.transformType(elemTy)(ctx)
+      val arrTyName = WasmTypeName.WasmArrayTypeName(IRTypes.ArrayType(typeRef))
+      ctx.addArrayType(
+        WasmArrayType(
+          arrTyName,
+          WasmStructField(Names.WasmFieldName.arrayField, wasmElemTy, isMutable = true)
+        )
+      )
+      if (lengths.tail.isEmpty) {
+        val zero = Defaults.defaultValue(wasmElemTy)
+        instrs += zero
+      } else {
+        genWithDimensions(
+          IRTypes.ArrayTypeRef(typeRef.base, typeRef.dimensions - 1),
+          lengths.tail
+        )
```

Comment:
IIUC, this will create a unique nested array and use it for *all* the elements of the outer array. In other words, it will compile
```scala
new Array[Array[Int]](2, 2)
```
into something equivalent to
```scala
val inner = new Array[Int](2)
val outer = Array(inner, inner)
```
But it is supposed to create a *different* instance of the inner array for each slot of the outer array.

I don't think it is possible to do this without a generic recursive helper, which we don't have enough infrastructure for at the moment.

I suggest focusing on the case where `lengths.size == 1` for now, and leave a `???` for this case. It is rather rare, so we should be able to make a lot of code work without that support.